src/libraries/System.Text.Json/src/ILLink/ILLink.Substitutions.xml

```diff
@@ -0,0 +1,8 @@
+<linker>
+  <assembly fullname="System.Text.Json">
+    <type fullname="System.Text.Json.JsonSerializer">
+      <method signature="System.Boolean get_IsReflectionEnabledByDefault()" body="stub" value="false" 
+              feature="System.Text.Json.JsonSerializer.IsReflectionEnabledByDefault" featurevalue="false"/>
```

Comment:
It sounds like ILLink.targets would be the best, but I don't have a strong opinion.

Example:
* We might want `JsonSerializer.IsReflectionEnabledByDefault` to default to false when trimming.
* But there will be SDKs that want it to default to true.
* At the same time, we want user to be able to still set it to false in their project.

So the scheme I'm thinking about is (I made up the property names):

* SDKs that want a different default set `<_IsJsonSerializerReflectionEnabledByDefault>true<_IsJsonSerializerReflectionEnabledByDefault>` in their .props
* In ILLink targets we set `<IsJsonSerializerReflectionEnabledByDefault Condition="$(IsJsonSerializerReflectionEnabledByDefault) == '' and $(_IsJsonSerializerReflectionEnabledByDefault) != 'true'">false<IsJsonSerializerReflectionEnabledByDefault>`

> BTW - how are these MSBuild settings tested? Are they tested in dotnet/runtime? or in dotnet/sdk?

For the `EventSourceSupport` one in Native.targets, we don't have a dedicated test, but there are libraries tests that set it in their csproj because they need EventSource and act as "positive tests", and then we have size testing in the perf repo that would detect the massive regression if this is accidentally enabled by default - that's the "negative test"...

Reply:
The publish targets in the sdk repo also seem relevant:

https://github.com/dotnet/sdk/blob/e0390ce4ef21db0205cf0b16ffb93a5b659c64b4/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L19-L22

Reply:
Have we decided we want this to be defaulted to `off` only when `PublishAot=true`? Or do we think we should go as far as turning it off by default when `PublishTrimmed=true`?

For the ASP.NET feature switch, [we had it for both](https://github.com/dotnet/sdk/blob/e0390ce4ef21db0205cf0b16ffb93a5b659c64b4/src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets#L51). But I'm a little concerned about making the switch for `PublishTrimmed` at the bottom of the stack. It is a breaking change. If we do this, we will need to update Maui and Blazor WASM SDKs to turn it back `on`.

Reply:
I agree that turning off reflection for `PublishTrimmed=true` might be overdoing it. OTOH turning it off for `PublishAot=true` feels like a sound default.

> We don't need to block this PR on the decision. 

Do you mean authoring the MSBuild property in general, or just linking it to other properties?

Reply:
I meant defaulting `JsonSerializerIsReflectionEnabledByDefault` based on `PublishXX`.

Reply:
https://github.com/dotnet/sdk/pull/31626

Reply:
> I agree that turning off reflection for PublishTrimmed=true might be overdoing it

Are there cases where the reflection fallback is 100% reliable and warning free with `PublishTrimmed=true`? As far as I know, the reflection fallback is always unreliable with System.Text.Json and trimming. We should not default to unreliable configurations with trimming.

Reply:
I'm not sure if we could ever call it reliable, but it is known to work in certain configurations which customers appear to depend on (it's a brittle arrangement of course, it takes one seeming unrelated change for their code to break). Re: warning-free the current state STJ APIs mean that you'd be getting false-positive warnings even in the case where source gen is being used, and this feature flag is precisely for regulating APIs that are annotated RUC/RDC.

Reply:
It is expected that the customers with brittle trimming arrangements may see breaks after upgrade and may apply workarounds to keep their app working with the brittle arrangement. This is for the product as a whole. It does not make sense to treat System.Text.Json differently, it just creates confusion.

Reply:
I've opened https://github.com/dotnet/runtime/issues/84378 to track this separately.